src/emulator/module/include/module/lay_out_args.h

```diff
@@ -98,17 +92,24 @@ constexpr void add_args_to_layout(ArgLayout &head, LayoutArgsState &state) {
 
     // Recursively add the remaining arguments.
     if (sizeof...(Tail) > 0) {
-        add_args_to_layout<Tail...>(*(&head + 1), state);
+        *(&state + 1) = state;
+        add_args_to_layout<Tail...>(*(&head + 1), *(&state + 1));
     }
 }
 
 template <typename... Args>
-constexpr ArgsLayout<Args...> lay_out() {
+constexpr std::tuple<ArgsLayout<Args...>, LayoutArgsState> lay_out() {
+    std::array<LayoutArgsState, sizeof...(Args)> states = {};
```

Comment:
Do we need an array of states? I think we should only need to return the last one, after all the (non-varargs) args have been added.

Reply:
I would also add a special case in `add_arg_to_layout` (hidden in diff so I can't comment) to not do anything to the state for varargs type.